File: src/main/java/com/babylonml/backend/training/optimizer/SimpleGradientDescentOptimizer.java
package com.babylonml.backend.training.optimizer;

import com.babylonml.backend.cpu.TensorOperations;
import com.babylonml.backend.training.execution.TrainingExecutionContext;
import com.babylonml.backend.training.execution.ContextInputSource;
import com.babylonml.backend.training.operations.MiniBatchListener;
import com.babylonml.backend.cpu.VectorOperations;
import com.babylonml.backend.training.operations.Operation;
import it.unimi.dsi.fastutil.ints.IntImmutableList;
import org.jspecify.annotations.NonNull;

import java.util.List;

public class SimpleGradientDescentOptimizer implements GradientOptimizer, MiniBatchListener {
    private int scaleValue = 1;

    public SimpleGradientDescentOptimizer(@NonNull ContextInputSource inputSource) {
        inputSource.addMiniBatchListener(this);
    }

    @Override
    public void optimize(@NonNull TrainingExecutionContext executionContext, float @NonNull [] matrix, int matrixOffset,
                         IntImmutableList shape, float @NonNull [] gradient, int gradientOffset, float learningRate, Operation operation) {
        var pointer = executionContext.allocateBackwardMemory(operation, shape);
        var buffer = pointer.buffer();
        var bufferOffset = pointer.offset();

        var stride = TensorOperations.stride(shape);
        VectorOperations.multiplyVectorToScalar(gradient, gradientOffset,
                -learningRate / scaleValue, buffer, bufferOffset,
                stride);
        VectorOperations.addVectorToVector(matrix, matrixOffset, buffer, bufferOffset, matrix, matrixOffset,
                stride);
    }

    @Override
    public @NonNull List<IntImmutableList> calculateRequiredMemoryAllocations(IntImmutableList shape) {
        return List.of(shape);
    }

    @Override
    public void onMiniBatchStart(long miniBatchIndex, int miniBatchSize) {
        assert miniBatchSize > 0;
        scaleValue = miniBatchSize;
    }
}

File: src/main/kotlin/com/babylonml/backend/training/optimizer/SimpleGradientDescentOptimizer.kt
package com.babylonml.backend.training.optimizer

import com.babylonml.backend.cpu.TensorOperations
import com.babylonml.backend.cpu.VectorOperations
import com.babylonml.backend.training.execution.ContextInputSource
import com.babylonml.backend.training.execution.TrainingExecutionContext
import com.babylonml.backend.training.operations.MiniBatchListener
import com.babylonml.backend.training.operations.Operation
import it.unimi.dsi.fastutil.ints.IntImmutableList

class SimpleGradientDescentOptimizer(inputSource: ContextInputSource) : GradientOptimizer, MiniBatchListener {
    private var scaleValue = 1

    init {
        inputSource.addMiniBatchListener(this)
    }

    override fun optimize(
        executionContext: TrainingExecutionContext,
        matrix: FloatArray,
        matrixOffset: Int,
        shape: IntImmutableList,
        gradient: FloatArray,
        gradientOffset: Int,
        learningRate: Float,
        operation: Operation
    ) {
        val pointer = executionContext.allocateBackwardMemory(operation, shape)
        val buffer = pointer.buffer()
        val bufferOffset = pointer.offset()

        val stride = TensorOperations.stride(shape)
        VectorOperations.multiplyVectorToScalar(
            gradient, gradientOffset,
            -learningRate / scaleValue, buffer, bufferOffset,
            stride
        )
        VectorOperations.addVectorToVector(
            matrix, matrixOffset, buffer, bufferOffset, matrix, matrixOffset,
            stride
        )
    }

    override fun calculateRequiredMemoryAllocations(shape: IntImmutableList): List<IntImmutableList> {
        return listOf(shape)
    }

    override fun onMiniBatchStart(miniBatchIndex: Long, miniBatchSize: Int) {
        assert(miniBatchSize > 0)
        scaleValue = miniBatchSize
    }
}
